impress-org/givewp-next-gen
# Feature: add support for multipart/form-data and file field uploads

Depends on https://github.com/impress-org/givewp/pull/6857, https://github.com/impress-org/givewp/pull/6859, https://github.com/impress-org/givewp/pull/6865

## Description



This replaces our main donation request to use `multipart/form-data` instead of `application/json` and handles storage of file fields.  This takes advantage of the web api [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData) interface to build the data for the request.  Since we are using the native [fetch](https://developer.mozilla.org/en-US/docs/Web/API/fetch) api it will automatically set the header content-type to `multipart/form-data` when using FormData for the body.

**Additional context**
Typical text-based forms (like our v2 forms) use `application/x-www-form-urlencoded` for content-type.
> In human terms, this means: "This is form data that has been encoded into URL parameters."

In v3 forms we originally opted to send our data as `application/json` really out of convenience and parity between our client and server because react-hook-form passes our handleSubmit callback a json object of values.

File uploads are considered a [special case](https://developer.mozilla.org/en-US/docs/Learn/Forms/Sending_and_retrieving_form_data#a_special_case_sending_files) because they are binary data.  This means the other available content-types won't be able to send this type of data to the server.  This is where the `multipart/form-data` content-type comes in and handles sending our form data as binary data.  The only real disadvantage I can see here is adding extra bytes to the request and a little less flexibility for nested data values.

Now, we _could_ have created a separate endpoint to handle file uploads specifically and left the form request _as is_ to use json but that would require a lot of overhead for one field type which would have to bypass our main donation endpoints.  In fact, that's basically how form field manager works.  However, this solution requires no overhead with a simple change to the request content-type.  It works seamlessly with our framework and and can still run through our validation process and along with the main donation request.  It also brings back native PHP superglobals to retreive data through `$_POST` and `$_FILES` where as `application/json` requires a raw input stream using `file_get_contents( 'php://input' )`.

## Affects



Since we're updating the form request content-type this will affect v3 form submission.

## Visuals



N/A

## Testing Instructions



Make sure GiveWP is `>=` https://github.com/impress-org/givewp/pull/6857 & https://github.com/impress-org/givewp/pull/6859 & https://github.com/impress-org/givewp/pull/6865

- First make sure v3 form is being processed properly
- Then programmatically add a file field using the snippet below and make sure the file gets uploaded to the media library and the donation meta records the media ID.

```
use Give\Framework\FieldsAPI\DonationForm;

add_action('givewp_donation_form_schema', static function (DonationForm $form) {
    $file = \Give\Framework\FieldsAPI\File::make('custom_file')
           ->maxUploadSize(wp_max_upload_size())
           ->allowedMimeTypes(get_allowed_mime_types())
            ->label('Custom File');

    $form->insertAfter('donationAmount', $file);
});
```

## Pre-review Checklist
